WinRT.Runtime/Projections.cs

```diff
@@ -247,6 +249,92 @@ public static bool TryGetCompatibleWindowsRuntimeTypeForVariantType(Type type, o
             return true;
         }
 
+        private static HashSet<Type> GetCompatibleTypes(Type type)
+        {
+            HashSet<Type> compatibleTypes = new HashSet<Type>();
+
+            foreach (var iface in type.GetInterfaces())
+            {
+                if (IsTypeWindowsRuntimeTypeNoArray(iface))
+                {
+                    compatibleTypes.Add(iface);
+                }
+
+                if (iface.IsConstructedGenericType
+                    && TryGetCompatibleWindowsRuntimeTypesForVariantType(iface, out var compatibleIfaces))
+                {
+                    compatibleTypes.UnionWith(compatibleIfaces);
+                }
+            }
+
+            Type baseType = type.BaseType;
+            while (baseType != null)
+            {
+                if (IsTypeWindowsRuntimeTypeNoArray(baseType))
+                {
+                    compatibleTypes.Add(baseType);
+                }
+                baseType = baseType.BaseType;
+            }
+
+            return compatibleTypes;
+        }
+
+        internal static bool TryGetCompatibleWindowsRuntimeTypesForVariantType(Type type, out IEnumerable<Type> compatibleTypes)
+        {
+            compatibleTypes = null;
+            if (!type.IsConstructedGenericType)
+            {
+                throw new ArgumentException(nameof(type));
+            }
+
+            var definition = type.GetGenericTypeDefinition();
+
+            if (!IsTypeWindowsRuntimeTypeNoArray(definition))
+            {
+                return false;
+            }
+
+            var genericConstraints = definition.GetGenericArguments();
+            var genericArguments = type.GetGenericArguments();
+            List<List<Type>> compatibleTypesPerGeneric = new List<List<Type>>();
+            for (int i = 0; i < genericArguments.Length; i++)
+            {
+                List<Type> compatibleTypesForGeneric = new List<Type>();
+                bool argumentCovariantObject = (genericConstraints[i].GenericParameterAttributes & GenericParameterAttributes.VarianceMask) == GenericParameterAttributes.Covariant
+                    && !genericArguments[i].IsValueType;
+
+                if (IsTypeWindowsRuntimeTypeNoArray(genericArguments[i]))
+                {
+                    compatibleTypesForGeneric.Add(genericArguments[i]);
+                }
+                else if (!argumentCovariantObject)
+                {
+                    return false;
+                }
+
+                if (argumentCovariantObject)
+                {
+                    compatibleTypesForGeneric.AddRange(GetCompatibleTypes(genericArguments[i]));
+                }
+
+                compatibleTypesPerGeneric.Add(compatibleTypesForGeneric);
+            }
+
+            // https://docs.microsoft.com/en-us/archive/blogs/ericlippert/computing-a-cartesian-product-with-linq

```

Comment:
hmm, this looks strangely familiar ;-)
https://github.com/microsoft/CsWinRT/compare/jkoritzinsky/multiple-PIIDs

Reply:
Yea it is basically the same, it is the non-linq version.  Moved to that based on the other comment about linq can cause high memory usage.